backend/module/bot/slackbot/slackbot.go

```diff
@@ -0,0 +1,169 @@
+	t := slackevents.TokenComparator{VerificationToken: token}
+	ok := t.Verify(requestToken)
+	if !ok {
+		return errors.New("invalid verification token")
+	}
+	return nil
+}
+
+// event type recieved the first time when configuring Slack API to send requests to our endpoint, https://api.slack.com/events/url_verification
+func (m *mod) handleURLVerificationEvent(challenge string) (*slackbotv1.EventResponse, error) {
+	// this should never happen
+	if challenge == "" {
+		msg := "slack API provided an empty challenge string"
+		m.logger.Error(msg)
+		return nil, status.Error(codes.InvalidArgument, msg)
+	}
+	return &slackbotv1.EventResponse{Challenge: challenge}, nil
+}
+
+// this outer event is 1 of 3 types: url_verification, event_callback, and app_rate_limited
+// this should be called via `go` in order to avoid blocking main execution
+func (m *mod) handleEvent(req *slackbotv1.EventRequest) {
+	switch req.Type {
+	case slackevents.CallbackEvent:
+		err := m.handleCallBackEvent(req.Event)
+		if err != nil {
+			m.logger.Error("handle callback event error", log.ErrorField(err))
+		}
+	case slackevents.AppRateLimited:
+		// received if endpoint got more than 30,000 request events in 60 minutes, https://api.slack.com/apis/connections/events-api#the-events-api__responding-to-events__rate-limiting
+		m.logger.Error("app's event subscriptions are being rate limited")
+	default:
+		m.logger.Error("received unexpected event type", zap.String("event type", req.Type))
+	}
+}
+
+// for the inner event, we currently support 2 types: app_mention (messages that mention the bot directly) and message (specifically DMs with the bot)
+// full list of Slack event types: https://api.slack.com/events
+func (m *mod) handleCallBackEvent(event *slackbotv1.Event) error {
+	switch event.Type {
+	case slackevents.AppMention:
+		return m.handleAppMentionEvent(event)
+	case slackevents.Message:
+		return m.handleMessageEvent(event)
+	default:
+		return fmt.Errorf("received unsuported event type: %s", event.Type)
+	}
+}
+
+// event type for messages that mention the bot directly
+func (m *mod) handleAppMentionEvent(event *slackbotv1.Event) error {
+	match := m.bot.MatchCommand(event.Text)
+	err := sendBotReply(m.slack, event.Channel, match)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// event type for DMs with the bot
+func (m *mod) handleMessageEvent(event *slackbotv1.Event) error {
+	if event.ChannelType != dmMessage {
+		return fmt.Errorf("unexpected channel type: %s", event.ChannelType)
+	}
+	// it's standard behavior of the Slack Events API that the bot will receive all message events, including from it's own posts
+	// so we need to filter out these events by checking that the BotId field is empty; otherwise by reacting to them we will create an infinte loop
+	if event.BotId == "" {
+		match := m.bot.MatchCommand(event.Text)
+		err := sendBotReply(m.slack, event.Channel, match)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
```

Comment:
nit try not to nest the happy path, it's more readable. here's a good [blog post](https://medium.com/@matryer/line-of-sight-in-code-186dd7cdea88) on the technique

```suggestion
	if event.BotId != "" {
		return nil
	}
	
	match := m.bot.MatchCommand(event.Text)
	return sendBotReply(m.slack, event.Channel, match)
```

Reply:
ty! fixed